### PR TITLE
Cleanup PARAFAC2 tolerances

### PR DESCRIPTION
**Pull request by aarmey**
Part of #481, this defines the tolerance in PARAFAC2 identically to the PARAFAC methods.

I still need to review the docstrings and explain a few more points justifying this change here.

### CODE REVIEW

**Comment by aarmey** [comment]
This is ready to review.

File: tensorly/decomposition/_parafac2.py [hunk 1]
```diff
@@ -573,11 +560,7 @@ def parafac_updates(X, w, f):
                         f"PARAFAC2 reconstruction error={rec_errors[-1]}, variation={rec_errors[-2] - rec_errors[-1]}."
                     )
 
-                if (
-                    tl.abs(rec_errors[-2] ** 2 - rec_errors[-1] ** 2)
-                    < (tol * rec_errors[-2] ** 2)
-                    or rec_errors[-1] ** 2 < absolute_tol
-                ):
+                if rec_errors[-2] - rec_errors[-1] < tol:
```
Comment (JeanKossaifi) [inline tensorly/decomposition/_parafac2.py:563]: Should there be an absolute value here?

File: tensorly/tenalg/svd.py [hunk 2]
```diff
@@ -357,6 +358,7 @@ def randomized_svd(
 
 
 SVD_FUNS = ["truncated_svd", "symeig_svd", "randomized_svd"]
+SVD_TYPES = Literal["truncated_svd", "symeig_svd", "randomized_svd"]
```
Comment (JeanKossaifi) [inline tensorly/tenalg/svd.py:361]: Should it just be Literal[f for f in SVD_FUNS] ?

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @aarmey sorry just saw you marked it as ready for review.

It looks good, just left some comments - I somehow find type hints more difficult to parse than just signature + doc, especially when it gets nested like Optional[Union[..]...] but I guess it can be nice to have for type check.

File: tensorly/tenalg/svd.py (hunk 2 above)
Comment (yngvem) [reply tensorly/tenalg/svd.py:361]: Unfortunately, `Literal` [doesn't allow arbitrary values](https://typing.readthedocs.io/en/latest/spec/literal.html#illegal-parameters-for-literal-at-type-check-time). A hacky workaround is

```
SVD_TYPES = Literal["truncated_svd", "symeig_svd", "randomized_svd"]
SVD_FUNS = list(SVD_TYPES.__args__)
```

but I'm not sure if that's any better than just duplicating the string, which makes it absolutely clear for the reader what's happening with the code.

**Comment by yngvem** [comment]
Great job @aarmey! I haven't done a super thorough review, but it looks great at first glance!

I have one question. I see you're using `Union`- and `Optional`-types instead of the new `|`- and ` | None`-syntax. I assume this is to support Python 3.9, which is [EOL next year](https://devguide.python.org/versions/)?

A workaround if we want to use the new syntax is adding a `from __future__ import annotations`, which turns all type annotations into strings. The upside is that it makes type annotations forward compatible, and a downside is that arbitrary syntax is allowed so typos might not be found. However IDEs are usually good at spotting typos, so I think it's worth it for the increased readibility.

Also, the previous default tolerance of `1e-8` was selected based on @MarieRoald's experience with PARAFAC2, so if we change the tolerance function, it might be worthwhile to also check that the default value is sensible (and works for non-simulated data also), but you might have done that already?

**Comment by aarmey** [comment]
> Thanks @aarmey sorry just saw you marked it as ready for review.
> 
> It looks good, just left some comments - I somehow find type hints more difficult to parse than just signature + doc, especially when it gets nested like Optional[Union[..]...] but I guess it can be nice to have for type check.

I agree that the type annotations can become unwieldy, @JeanKossaifi. I have found them to still be helpful, on balance, because they describe the argument types in a very exact manner:

1. Occasionally, the documentation is a little vague because arguments such as `nn_mode` are quite complicated. A more advanced user can inspect these to know exactly what can be passed in.
2. These seem to help with autocomplete tools, probably because they are machine-readable.
3. We have a complicated codebase, and I've found automated checkers such as ruff/pyright to be helpful. For instance, quickly skimming the results, `tl.linalg.higher_order_moment()` currently does not work because `batched_outer()` only takes one argument, not two. It's easy to overdo it with automated tools, but I wonder if these plus some type annotation could help to avoid a separate set of errors compared to our tests. (I have little experience with large codebases, so I am genuinely wondering what people think about this.)

**Comment by aarmey** [comment]
Thanks, @JeanKossaifi and @yngvem! I like this plan, so I'll merge this and then follow up with a PR removing Python 3.9 support and simplifying the type annotations.

**Pull request merged**